sql/memo/coster.go

```diff
@@ -171,8 +171,96 @@ func (c *coster) costHashJoin(ctx *sql.Context, n *HashJoin, _ sql.StatsReader)
 }
 
 func (c *coster) costMergeJoin(_ *sql.Context, n *MergeJoin, _ sql.StatsReader) (float64, error) {
+
 	l := n.Left.RelProps.card
-	return l * cpuCostFactor, nil
+	r := n.Right.RelProps.card
+
+	comparer, ok := n.Filter[0].(*Equal)
+	if !ok {
+		return 0, sql.ErrMergeJoinExpectsComparerFilters.New(n.Filter[0])
+	}
+
+	var leftCompareExprs []*ExprGroup
+	var rightCompareExprs []*ExprGroup
+
+	leftTuple, isTuple := comparer.Left.Scalar.(*Tuple)
+	if isTuple {
+		rightTuple, _ := comparer.Right.Scalar.(*Tuple)
+		leftCompareExprs = leftTuple.Values
+		rightCompareExprs = rightTuple.Values
+	} else {
+		leftCompareExprs = []*ExprGroup{comparer.Left}
+		rightCompareExprs = []*ExprGroup{comparer.Right}
+	}
+
+	if isLinearMerge(n, leftCompareExprs, rightCompareExprs) {
+		// We're guarenteed that the execution will never need to iterate over multiple rows in memory.
+		return (l + r) * cpuCostFactor, nil
+	}
+
+	// Each comparison either compares a row on one child table with a constant, which reduces the number of rows from
+	// that child, or compares a column on each table, which reduces the expected number of collisions on the comparator.
+	// In either case, we assume a similar improvement in the selectivity of the join.
+	selectivity := math.Pow(perKeyCostReductionFactor, float64(len(leftCompareExprs)))
+	return (l + r + l*r*selectivity) * cpuCostFactor, nil
+}
+
+// isLinearMerge determines whether either of a merge join's child indexes returns only unique values for the merge
+// comparator.
+func isLinearMerge(n *MergeJoin, leftCompareExprs, rightCompareExprs []*ExprGroup) bool {
```

Comment:
double check me on this, but the property here is maybe "isInjectiveMerge?" f(x1) == f(x2) implies x1 == x2

Reply:
Done.